sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchemaTransformWriteProvider.java

```diff
@@ -0,0 +1,173 @@
+            @UnknownKeyFor @NonNull @Initialized PCollectionRowTuple,
+            @UnknownKeyFor @NonNull @Initialized PCollectionRowTuple>
+        buildTransform() {
+      return new PTransform<@NonNull PCollectionRowTuple, @NonNull PCollectionRowTuple>() {
+        @Override
+        public PCollectionRowTuple expand(@NonNull PCollectionRowTuple input) {
+          SpannerWriteResult result =
+              input
+                  .get("input")
+                  .apply(
+                      MapElements.via(
+                          new SimpleFunction<Row, Mutation>(
+                              row ->
+                                  MutationUtils.createMutationFromBeamRows(
+                                      Mutation.newInsertOrUpdateBuilder(configuration.getTableId()),
+                                      Objects.requireNonNull(row))) {}))
+                  .apply(
+                      SpannerIO.write()
+                          .withDatabaseId(configuration.getDatabaseId())
+                          .withInstanceId(configuration.getInstanceId())
+                          .withFailureMode(SpannerIO.FailureMode.REPORT_FAILURES));
+          Schema failureSchema =
+              Schema.builder()
+                  .addStringField("operation")
+                  .addStringField("instanceId")
+                  .addStringField("databaseId")
+                  .addStringField("tableId")
+                  .addStringField("mutationData")
+                  .build();
+          result
+              .getFailedMutations()
+              .apply(
+                  FlatMapElements.into(TypeDescriptors.rows())
+                      .via(
+                          mtg ->
+                              Objects.requireNonNull(mtg).attached().stream()
+                                  .map(
+                                      mutation ->
+                                          Row.withSchema(failureSchema)
+                                              .addValue(mutation.getOperation().toString())
+                                              .addValue(configuration.getInstanceId())
+                                              .addValue(configuration.getDatabaseId())
+                                              .addValue(mutation.getTable())
+                                              // TODO(pabloem): Figure out how to represent
+                                              // mutation
+                                              //  contents in DLQ
+                                              .addValue(
+                                                  Iterators.toString(
+                                                      mutation.getValues().iterator()))
+                                              .build())
+                                  .collect(Collectors.toList())))
+              .setRowSchema(failureSchema);
+          return PCollectionRowTuple.empty(input.getPipeline());
```

Comment:
Is there any way to convert the mutations back to rows at this stage and return the failed rows?

Reply:
also planning to leave this work for later as we productionize the transform. is that ok?

Reply:
Yup! can you leave a TODO for later? maybe someone can pick it up. Same with the validation checks

Reply:
done and added a rudimentary return value for now